openrouteservice/src/main/java/org/heigit/ors/matrix/algorithms/core/CoreMatrixAlgorithm.java

```diff
@@ -0,0 +1,548 @@
+import com.graphhopper.routing.weighting.Weighting;
+import com.graphhopper.storage.CHGraph;
+import com.graphhopper.storage.Graph;
+import com.graphhopper.util.EdgeExplorer;
+import com.graphhopper.util.EdgeIterator;
+import com.graphhopper.util.EdgeIteratorState;
+import org.heigit.ors.matrix.*;
+import org.heigit.ors.matrix.algorithms.AbstractMatrixAlgorithm;
+import org.heigit.ors.matrix.algorithms.dijkstra.DijkstraManyToMany;
+import org.heigit.ors.routing.algorithms.SubGraph;
+import org.heigit.ors.routing.graphhopper.extensions.core.CoreDijkstraFilter;
+import org.heigit.ors.routing.graphhopper.extensions.core.CoreMatrixFilter;
+import org.heigit.ors.routing.graphhopper.extensions.edgefilters.EdgeFilterSequence;
+import org.heigit.ors.routing.graphhopper.extensions.storages.AveragedMultiTreeSPEntry;
+import org.heigit.ors.routing.graphhopper.extensions.storages.MultiTreeSPEntryItem;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.PriorityQueue;
+
+/**
+ * A Core and Dijkstra based algorithm that calculates the weights from multiple start to multiple goal nodes.
+ * Using core and true many to many.
+ * @author Hendrik Leuschner
+ */
+public class CoreMatrixAlgorithm extends AbstractMatrixAlgorithm {
+    protected int coreNodeLevel;
+    protected int maxVisitedNodes = Integer.MAX_VALUE;
+    protected int visitedNodes;
+    private int treeEntrySize;
+    private boolean hasTurnWeighting = false;
+    private boolean swap = false;
+
+    private PriorityQueue<AveragedMultiTreeSPEntry> upwardQueue;
+    private IntHashSet coreEntryPoints;
+    private IntHashSet coreExitPoints;
+    private IntObjectMap<AveragedMultiTreeSPEntry> bestWeightMap;
+    private IntObjectMap<List<AveragedMultiTreeSPEntry>> bestWeightMapCore;
+    private IntObjectMap<AveragedMultiTreeSPEntry> targetMap;
+    private IntHashSet targetSet;
+    private MultiTreeMetricsExtractor pathMetricsExtractor;
+    private CoreDijkstraFilter additionalCoreEdgeFilter;
+    private CHGraph chGraph;
+    private SubGraph targetGraph;
+    private TurnWeighting turnWeighting;
+
+    @Override
+    public void init(MatrixRequest req, GraphHopper gh, Graph graph, FlagEncoder encoder, Weighting weighting) {
+        if (weighting instanceof TurnWeighting) {
+            hasTurnWeighting = true;
+            turnWeighting = (TurnWeighting) weighting;
+        }
+        weighting = new PreparationWeighting(weighting);
+        super.init(req, gh, graph, encoder, weighting);
+        try {
+            chGraph = graph instanceof CHGraph ? (CHGraph) graph : (CHGraph) ((QueryGraph) graph).getMainGraph();
+        } catch (ClassCastException e) {
+            throw new ClassCastException(e.getMessage());
+        }
+        coreNodeLevel = chGraph.getNodes() + 1;
+        pathMetricsExtractor = new MultiTreeMetricsExtractor(req.getMetrics(), graph, this.encoder, weighting, req.getUnits());
+        initCollections(10);
+        initFilter(null, chGraph);
```

Comment:
Is there any particular reason why `initFilter` method is called at least twice: first here, but then being overridden by the second call in the upstream `init` methods in lines 95 and 100?

Reply:
You're right, one of them can be removed. Those are different init functions under different circumstances (1. to adhere to `AbstractMatrixAlgorithm`, then to init with a filter, then to init without a GH instance). One of the initFilters has been removed.

Reply:
Thanks for simplifying the init methods! While I understand the objective behind adhering to the upstream method, I still think the design could be improved by avoiding calling `initFilter` twice. In fact, I don't even think there is real need for this method. For example, the filter could be initialized directly in L90 by having `additionalCoreEdgeFilter = new CoreMatrixFilter(chGraph);` while the call to `initFilter` in L95 could be replaced by `additionalCoreEdgeFilter.addRestrictionFilter(additionalEdgeFilter);`. Don't you think this would be simpler and easier to follow?

Reply:
No, I think separating mulit-line functionality out into a method with a descriptive name such as `initFilter` is a better way to go about it.

Reply:
Thanks for explaining your point of view. I'm afraid that I have to disagree. First of all, this is not multi-line functionality. It appears so only because `initFilter` is called twice each time with different purpose and needs additional logic to handle these two scenarios. In fact, if you would remove `initFilter` each previous call to it could be replaced by a single line of code, leading to an overall reduction of code. The check `additionalEdgeFilter != null` is actually not necessary because of L82 in `CoreDijkstraFilter`. I agree that it makes sense to abstract one-liners to separate methods with descriptive name but only if these are called more than once. This is not the case here. Having said all this, the current implementation seems over-engineered to me and actually bad design because the same code has to executed twice while it doesn't have to.

Reply:
sure. done.